src/sparse_merkle_tree/smt.rs

```diff
@@ -1556,6 +1557,137 @@ impl SparseMerkleTree {
+                    return Err(SMTError::InvalidInput(String::from(
+                        "Not all sibling hashes were used",
+                    )));
+                }
+                for hash in adding_sibling_hashes {
+                    if base_sibling_hashes.contains(&hash) {
+                        return Err(SMTError::InvalidInput(String::from(
+                            "Duplicate sibling hashes",
+                        )));
+                    }
+                    base_sibling_hashes.push(hash);
+                }
+                for hash in removing_sibling_hashes {
+                    if let Some(pos) = base_sibling_hashes.iter().position(|x| x == &hash) {
+                        base_sibling_hashes.remove(pos);
+                    } else {
+                        return Err(SMTError::InvalidInput(String::from(
+                            "Removed key is not in sibling hashes",
+                        )));
+                    }
+                }
+                let mut updated_queries = proof.queries.clone();
+                for (index, proof_query) in proof.queries.iter().enumerate() {
+                    if removing_keys.contains(&proof_query.key()) {
+                        updated_queries.remove(index);
+                    }
+                }
+                proof.queries = updated_queries;
+                return Ok(Arc::from(Mutex::new(proof.clone())));
+            }
+
+            let mut sibling_hash: Vec<u8> = vec![];
+            // there are three cases for the sibling hash:
+            if !sorted_queries.is_empty() && query.is_sibling_of(&sorted_queries[0]) {
+                // #1. sibling hash is next element of sorted_queries.
+                let is_first_query_removed =
+                    removing_keys.contains(&sorted_queries[0].query_proof.key());
+                let sibling = sorted_queries.pop_front().unwrap();
+                // We are merging two branches.
+                // Check that the bitmap at the merging point is consistent with the nodes type.
+                let is_sibling_empty = utils::is_empty_hash(&sibling.hash);
+                if (is_sibling_empty && query.binary_bitmap[0])
+                    || (!is_sibling_empty && !query.binary_bitmap[0])
+                {
+                    return Err(SMTError::InvalidInput(String::from(
+                        "Bitmap is not consistent with the nodes type",
+                    )));
+                }
+                let is_query_empty = utils::is_empty_hash(&query.hash);
+                if (is_query_empty && sibling.binary_bitmap[0])
+                    || (!is_query_empty && !sibling.binary_bitmap[0])
+                {
+                    return Err(SMTError::InvalidInput(String::from(
+                        "Bitmap is not consistent with the nodes type",
+                    )));
+                }
+                if !utils::array_equal_bool(&query.binary_bitmap[1..], &sibling.binary_bitmap[1..])
```

Comment:
shoulnd't we check length of binary_bitmap here to ensure index 1 is accessible?

Reply:
we will check the length of the bitmap [here](https://github.com/LiskHQ/lisk-db/blob/9f6b634d621326aaa30e703b449d6bb16c9c4027/src/sparse_merkle_tree/smt.rs#L1651).